drivers/crc/crc_stm32.c

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (c) 2024 Brill Power Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+#include <errno.h>
+#include <soc.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+#include <zephyr/init.h>
+#include "stm32_ll_crc.h"
+#include "crc_hw.h"
+
+LOG_MODULE_REGISTER(crc_stm32);
+
+#define DT_DRV_COMPAT st_stm32_crc
+
+struct crc_cfg {
+	CRC_TypeDef *base;
+	struct stm32_pclken pclken;
+};
+
+static int crc_stm32_setup_crc_8_ccitt(const struct device *dev)
+{
+	const struct crc_cfg *cfg = dev->config;
+
+	LL_CRC_SetInputDataReverseMode(cfg->base, LL_CRC_INDATA_REVERSE_NONE);
+	LL_CRC_SetOutputDataReverseMode(cfg->base, LL_CRC_OUTDATA_REVERSE_NONE);
+	LL_CRC_ResetCRCCalculationUnit(cfg->base);
+	LL_CRC_SetPolynomialSize(cfg->base, LL_CRC_POLYLENGTH_8B);
+	LL_CRC_SetPolynomialCoef(cfg->base, 0x07);
+
+	return 0;
+}
+
+static uint8_t crc_stm32_crc_8_ccitt(const struct device *dev, uint8_t initial_value,
+									 const void *buf, size_t len)
+{
+	const struct crc_cfg *cfg = dev->config;
+	const uint8_t *num_data = buf;
+
+	LL_CRC_SetInitialData(cfg->base, initial_value);
+
+	register uint32_t index = 0;
+
+	for (index = 0; index < len; index++) {
+		LL_CRC_FeedData8(cfg->base, num_data[index]);
+	}
+
+	return LL_CRC_ReadData8(cfg->base);
+}
+
+static inline uint8_t crc_stm32_verify_crc_8_ccitt(const struct device *dev, uint8_t expected,
+													uint8_t initial_value, const void *buf,
+													size_t len)
+{
+	return (expected == crc_stm32_crc_8_ccitt(dev, initial_value, buf, len)) ? 1 : 0;
+}
+
+static int crc_stm32_setup_crc_32_ieee(const struct device *dev)
+{
+	const struct crc_cfg *cfg = dev->config;
+
+	LL_CRC_SetInputDataReverseMode(cfg->base, LL_CRC_INDATA_REVERSE_WORD);
+	LL_CRC_SetOutputDataReverseMode(cfg->base, LL_CRC_OUTDATA_REVERSE_BIT);
+	LL_CRC_ResetCRCCalculationUnit(cfg->base);
+	LL_CRC_SetPolynomialSize(cfg->base, LL_CRC_POLYLENGTH_32B);
+	LL_CRC_SetPolynomialCoef(cfg->base, LL_CRC_DEFAULT_CRC32_POLY);
+	LL_CRC_SetInitialData(cfg->base, LL_CRC_DEFAULT_CRC_INITVALUE);
+
+	return 0;
+}
+
+static uint32_t crc_stm32_crc_32_ieee(const struct device *dev, const void *buf, size_t len)
+{
+	const struct crc_cfg *cfg = dev->config;
+	const uint8_t *num_data = buf;
+
+	register uint32_t data = 0;
+	register uint32_t index = 0;
+
+	/* Compute the CRC of Data Buffer array */
+	for (index = 0; index < (len / 4); index++) {
+		data = (uint32_t)((num_data[4 * index + 3] << 24) | (num_data[4 * index + 2] << 16)
+							(num_data[4 * index + 1] << 8) | num_data[4 * index]);
```

Comment:
I think `sys_get_be32` from `<zephyr/sys/byteorder.h>` should be used instead.
```suggestion
		data = sys_get_be32(num_data);
```